sips/sip-011/sip-011-catastrophic-failure-recovery.md

```diff
@@ -0,0 +1,464 @@
+the blockchain from making progress.  For example, a bug that causes all nodes
+in the network to simultaneously crash is a liveness failure.  As another
+example, a bug that causes all nodes in the network to become unable to respond
+to network requests is a liveness failure.
+
+The second class of bugs are _safety failures_.  These are bugs in which the
+blockchain itself permits state-transitions that are undesirable and
+inconsistent with the design of the system.  For example, a bug that would let
+anyone move anyone else's STX tokens would be a safety failure.  As another
+example, a bug in which some Clarity code's execution led to the wrong result
+(but consistent across the network) would be a safety failure.
+
+This document describes the ways in which Stacks ecosystem participants can
+coordinate to recover from various instances of these failure modes, and when it
+is appropriate to use them.  These procedures include:
+
+* Announcing a new point-release of the Foundation's reference implementation of
+  the Stacks node software
+* Covertly fixing systemically-important nodes to fix security-sensitive bugs
+* Altering or extending the p2p network protocol
+* Soft forks, which constrain what kinds of transactions may be mined on the
+  canonical fork
+* Hard forks, which are backwards-incompatible upgrades
+
+## A Note on Scope
+
+The absence of a feature is not a catastrophic failure.  The blockchain behaving
+consistently with the SIPs that describe its design is not a catastrophic
+failure, even if the behavior is not desired.  This SIP does not describe how to
+remedy these situations; it is only concerned with manual recovery procedures
+and guidelines for restoring the network after a network-wide safety or liveness
+failure is discovered.
+
+# Specification
+
+The following catastrophic failure recovery procedures are presented in order
+from least-disruptive to most-disruptive.  When choosing the recovery procedure
+to execute, ecosystem participants should err on choosing the least disruptive
+procedure that will fix the problem at hand.  A recovery procedure should only
+be taken only once all less-disruptive recovery procedures have been considered
+or tried.
+
+## Procedure for Pushing out a Critical Bugfix Point-Release
+
+In the event that a safety or liveness failure can be readily addressed in
+public _without_ altering the blockchain's social contract (i.e. no changes to
+the networking protocols, mining protocols, or chainstate validation protocols),
+the following procedure will be carried out:
+
+1. The code for the fix will be written and pushed to a Github branch prefixed
+with `fix/` to the Stacks Blockchain reference implementation, hosted at
+https://github.com/blockstack/stacks-blockchain.
+2. The branch will be submitted as a pull-request to the `master` branch, and
+will be reviewed and approved by at least two blockchain engineers, representing
```

Comment:
[nit] s/"blockchain engineers"/"developers with write access to the stacks-blockchain repo"

Alternatively: "two members of the Stacks Core Developers group"?

Reply:
> In particular, is there some specific advantage an "unofficial" announcement has? If Github releases are a source of truth, availability of a new release is automatically a "formal" announcement IMO.

The point is to let node operators know when the source code for the fix is available for scrutiny and testing (which is mentioned at the end of (3)).

Reply:
Ah, I see Github rendered the quoted text in the above comment.

I don't think the set of people who have write access to the stacks blockchain repo accurately reflects the set of people will be called up on to fix catastrophic bugs.  I addressed this in 3832ae2 by defining the Stacks Core Developers as a list of developer names and contacts in a supplementary file, which can be added to or removed from by the Steering Committee.  To be clear, this is just the list of folks who will be called upon to execute these procedures; this isn't meant to be an exclusionary list of who's-who in the blockchain or some other social club or clique.  It's more like a "who is it okay to email at 3am on a Saturday if the blockchain halts and catches fire" list.